packages/core/lib/commands/migrate/runMigrations.js

```diff
@@ -1,6 +1,106 @@
+const { createReadStream } = require("fs");
+const { readdir } = require("fs/promises");
+const path = require("path");
+const JSONStream = require("JSONStream");
 const Migrate = require("@truffle/migrate").default;
+const TruffleError = require("@truffle/error");
+const os = require("os");
+const debug = require("debug")("migrate:run");
+
+// Search for Push1 (60) to Push(32) 7F + console.log
+//                      60 ---- 7F  C O N S O L E . L O G
+const consoleLogRex = /[67][0-9a-f]636F6e736F6c652e6c6f67/i;
+
+async function usesConsoleLog(artifactJson) {
+  const debugLog = debug.extend("test");
+  debugLog("Artifact: %o", artifactJson);
+
+  //create a parser to get the value of jsonpath .deployedBytecode
+  const parser = JSONStream.parse(["deployedBytecode"]);
+  const stream = createReadStream(artifactJson).pipe(parser);
+
+  return new Promise((resolve, reject) => {
+    stream.on("data", data => {
+      //JSONParse will emit the entire string/value
+      //so initiate stream cleanup here
+      stream.destroy();
+      const usesConsoleLog = consoleLogRex.test(data);
+      debugLog("usesConsoleLog:", usesConsoleLog);
+      resolve(usesConsoleLog);
+    });
+
+    stream.on("error", err => {
+      stream.destroy();
+      debugLog("onError: %o", err);
+      reject(err);
+    });
+  });
+}
+
+async function findArtifactsThatUseConsoleLog(buildDir) {
+  const debugLog = debug.extend("dirty-files");
+  const files = await readdir(buildDir);
+
+  let promises = [];
+  files.forEach(fn => {
+    if (fn.endsWith(".json")) {
+      promises.push(usesConsoleLog(path.join(buildDir, fn)));
+    }
+  });
+
+  const settled = await Promise.allSettled(promises);
+  const dirtyArtifacts = [];
+  for (const s of settled) {
+    if (s.status === "rejected") {
+      debugLog("promise failure: %o", s.reason);
+      continue;
+    }
+    dirtyArtifacts.push(s.value);
+  }
+  return dirtyArtifacts;
```

Comment:
This is ever so slightly more performant, as it avoids looping over the files twice. But there is an actual _fix_ here, in think, which is that `dirtyArtifacts` now contains the actual file names of the files that contain console.log, instead of `true`/`false`:

I didn't test this lol

```
  const dirtyArtifacts = [];
  await Promise.allSettled(files.map(async fn => {
    if (fn.endsWith(".json")) {
      try {
        const usedConsoleLog = await usesConsoleLog(path.join(buildDir, fn)));
        if (usedConsoleLog) {
          dirtyArtifacts.push(fn);
        }
      } catch(e) {
        debugLog("promise failure: %o", e.message);
      }
    }
  }));

  return dirtyArtifacts;
```

Reply:
p.s., please add a test for this :-)

Reply:
nice.